chain/client/src/chunks_delay_tracker.rs

```diff
@@ -1,192 +1,129 @@
-    heights: BTreeMap<BlockHeight, HeightInfo>,
+    /// Timestamp of trying to process a block for the first time.
+    blocks_received: HashMap<CryptoHash, Instant>,
+    chunks_in_progress: LruCache<ChunkHash, ChunkInProgress>,
 }
 
-const CHUNKS_DELAY_TRACKER_HORIZON: u64 = 10;
+const CHUNK_DELAY_CACHE_SIZE: usize = 1_000_000;
+
+impl Default for ChunksDelayTracker {
+    fn default() -> Self {
+        Self {
+            blocks_received: HashMap::new(),
+            chunks_in_progress: LruCache::new(CHUNK_DELAY_CACHE_SIZE),
+        }
+    }
+}
 
 #[derive(Debug, Default)]
-struct HeightInfo {
-    block_received: Option<Instant>,
-    chunks_received: HashMap<ShardId, Instant>,
+struct ChunkInProgress {
+    /// Timestamps of receiving each of the shards in a block.
+    chunk_received: Option<Instant>,
+    /// Timestamps of requesting missing chunks.
+    /// Chunks are usually requested when a client falls out-of-sync with the head of the chain.
+    chunk_requested: Option<Instant>,
 }
 
 impl ChunksDelayTracker {
-    // Blocks below this heights are not tracked.
-    fn lowest_height(head_height: BlockHeight) -> BlockHeight {
-        head_height.saturating_sub(CHUNKS_DELAY_TRACKER_HORIZON)
+    pub fn add_block_timestamp(&mut self, block_hash: &CryptoHash, timestamp: Instant) {
+        self.blocks_received.entry(*block_hash).or_insert(timestamp);
     }
 
-    fn remove_old_entries(&mut self, head_height: BlockHeight) {
-        let heights_to_drop: Vec<BlockHeight> = self
-            .heights
-            .range(..ChunksDelayTracker::lowest_height(head_height))
-            .map(|(&k, _v)| k)
-            .collect();
-        for h in heights_to_drop {
-            self.heights.remove(&h);
-        }
-    }
-
-    // Computes the delay between receiving a block and receiving the latest of its chunks.
-    // Updates the metric to the max delay across the most recently processed blocks.
-    fn get_max_delay(&mut self, head_height: BlockHeight) -> Duration {
-        let mut max_delay = Duration::ZERO;
-        for (_, v) in self.heights.range(..=head_height) {
-            if let Some(block_received) = v.block_received {
-                if let Some(latest_chunk) = v.chunks_received.values().max() {
-                    if latest_chunk > &block_received {
-                        let delay = latest_chunk.duration_since(block_received);
-                        max_delay = max(max_delay, delay);
-                    }
-                }
+    pub fn add_chunk_timestamp(&mut self, chunk_hash: &ChunkHash, timestamp: Instant) {
+        match self.chunks_in_progress.get_mut(chunk_hash) {
+            Some(entry) => {
+                entry.chunk_received.get_or_insert(timestamp);
+            }
+            None => {
+                self.chunks_in_progress.put(
+                    chunk_hash.clone(),
+                    ChunkInProgress { chunk_received: Some(timestamp), chunk_requested: None },
```

Comment:
It's ok to not track chunks that are not requested since we care about the delay between when a chunk is requested vs when it is received. If it is received before it is requested, we can just ignore this data. This also simplified the implementation so you can get rid of the None case here.

Reply:
Agreed